### PR TITLE
Add SCRAM-SHA-256 authentication support

### LINKED ISSUE #3784: Support `SCRAM-SHA-256` authentication mechanism
### What should be done?

We should implement `SCRAM-SHA-256` authentication mechanism. Credentials should be stored and validated on login. Enforcing authentication is in #1877.

The research prototype is in #3904, but its code should not be used. And we probably need to figure out integration testing first: both auth mechanisms (`PLAIN` and `SCRAM-SHA-256`) and new authentication enabled ~and disabled~.

### Where?

Handler.

### Definition of Done

- integration/compatibility tests added/updated;
- handler updated;
- spot refactorings done.

### PR DESCRIPTION
**Pull request by henvic**
# Description

We want to support SCRAM-SHA-256 as an authentication method.

Closes #3784.

## Readiness checklist

- [x] I added/updated unit tests (and they pass).
- [x] I added/updated integration/compatibility tests (and they pass).
- [ ] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [x] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

File: integration/users/connection_test.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+			username:       "scramsha256updated",
+			password:       "pass123",
+			updatePassword: "anotherpassword",
+			mechanism:      "SCRAM-SHA-256",
+		},
+		"BadPassword": {
+			username:      "baduser",
+			password:      "something",
+			mechanism:     "SCRAM-SHA-256",
+			wrongPassword: true,
```
Comment (rumyantseva) [inline integration/users/connection_test.go:70]: Do we test a wrong password for an existing user too?

**Comment by Neustradamus** [comment]
@henvic: Nice, good job!

Linked to:
- https://github.com/scram-sasl/info/issues/1
- https://github.com/FerretDB/FerretDB/issues/2012
- https://github.com/FerretDB/FerretDB/issues/3784
- https://github.com/FerretDB/FerretDB/pull/3904
- https://github.com/FerretDB/FerretDB/issues/3938
- https://github.com/FerretDB/FerretDB/pull/3970
- https://github.com/FerretDB/FerretDB/pull/3989

File: internal/util/password/scramsha256.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		return nil, err
+	}
+
+	salt := make([]byte, 45)
+	if _, err := rand.Read(salt); err != nil {
+		return nil, fmt.Errorf("failed to generate salt: %w", err)
+	}
+
+	kf := scram.KeyFactors{
+		Salt:  base64.StdEncoding.EncodeToString(salt),
```
Comment (b1ron) [inline internal/util/password/scramsha256.go:41]: That should not be [encoded](https://pkg.go.dev/github.com/xdg-go/scram@v1.1.2#KeyFactors).

```suggestion
		Salt:  string(salt),
```

That will fix the invalid client [proof](https://github.com/xdg-go/scram/blob/v1.1.2/server_conv.go#L143).

File: internal/util/password/scramsha256.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+		Iters: 15000,
+	}
+
+	cred := client.GetStoredCredentials(kf)
+
+	doc := must.NotFail(types.NewDocument())
+	doc.Set("storedKey", base64.StdEncoding.EncodeToString(cred.StoredKey))
+	doc.Set("iterationCount", int32(kf.Iters))
+	doc.Set("salt", kf.Salt)
```
Comment (b1ron) [inline internal/util/password/scramsha256.go:50]:
```suggestion
	doc.Set("salt", base64.StdEncoding.EncodeToString(salt))
```

File: internal/util/password/scramsha256.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	"encoding/base64"
+	"fmt"
+
+	"github.com/xdg-go/scram"
+
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+)
+
+// NewSCRAMSHA256 authentication.
```
Comment (b1ron) [inline internal/util/password/scramsha256.go:28]:
```suggestion
// NewSCRAMSHA256 computes the SCRAM-SHA-256 credentials and returns the document that should be stored.
```

File: internal/util/password/scramsha256.go (hunk 2 above)
Comment (AlekSi) [reply internal/util/password/scramsha256.go:41]: Without actually looking into it, may I suggest checking if `string` BSON type is not confused with `types.Binary`?

File: integration/users/connection_test.go (hunk 1 above)
Comment (henvic) [reply integration/users/connection_test.go:74]: Yes. This tested user actually exists.

**Review by noisersup — DISMISSED** [review]
Looks safe to me :)

File: internal/handler/msg_saslstart.go [hunk 5]
```diff
@@ -129,3 +147,117 @@ func saslStartPlain(doc *types.Document) (string, string, error) {
...
+
+	dbName, err := common.GetRequiredParam[string](doc, "$db")
+	if err != nil {
+		return "", err
+	}
+
+	scramServer, err := scram.SHA256.NewServer(func(username string) (scram.StoredCredentials, error) {
+		return h.credentialLookup(ctx, username, dbName)
+	})
+	must.NoError(err)
```
Comment (chilagrow) [inline internal/handler/msg_saslstart.go:251]: Should we handle error gracefully rather than panic?
```suggestion
	if err != nil{
		return "", err
	}
```

**Review by chilagrow — COMMENTED** [review]
Great work a few minor questions and comments :hugs:

File: internal/util/password/scramsha256.go
Comment (AlekSi) [inline internal/util/password/scramsha256.go:1]: Please extract changes in this package into a separate PR and make them closer to what we already have there, with a similar API

File: internal/util/password/scramsha256.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	"fmt"
+
+	"github.com/xdg-go/scram"
+
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+)
+
+// NewSCRAMSHA256 computes SCRAM-SHA-256 credentials and returns the document that should be stored.
+func NewSCRAMSHA256(username, password string) (*types.Document, error) {
```
Comment (AlekSi) [inline internal/util/password/scramsha256.go:29]:
```suggestion
func SCRAMSHA256Hash(username, password string) (*types.Document, error) {
```

File: internal/util/password/scramsha256.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	client, err := scram.SHA256.NewClient(username, password, "")
+	if err != nil {
+		return nil, err
+	}
+
+	salt := make([]byte, 45)
```
Comment (AlekSi) [inline internal/util/password/scramsha256.go:35]: That constant should be in the `fixedSCRAMSHA256Params` package-level variable

File: internal/util/password/scramsha256.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+	if _, err := rand.Read(salt); err != nil {
+		return nil, fmt.Errorf("failed to generate salt: %w", err)
+	}
+
+	kf := scram.KeyFactors{
+		Salt:  string(salt),
+		Iters: 15000,
```
Comment (AlekSi) [inline internal/util/password/scramsha256.go:42]: The same

File: internal/util/password/scramsha256.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	}
+
+	cred := client.GetStoredCredentials(kf)
+
+	doc := must.NotFail(types.NewDocument())
+	doc.Set("storedKey", base64.StdEncoding.EncodeToString(cred.StoredKey))
+	doc.Set("iterationCount", int32(kf.Iters))
+	doc.Set("salt", base64.StdEncoding.EncodeToString(salt))
+	doc.Set("serverKey", base64.StdEncoding.EncodeToString(cred.ServerKey))
```
Comment (AlekSi) [inline internal/util/password/scramsha256.go:51]: Are you sure that MongoDB stores those values as base64-encoded strings and not in types.Binary?

File: internal/util/password/scramsha256_test.go [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+	for name, tc := range tests {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			got, err := NewSCRAMSHA256(tc.username, tc.password)
+			require.NoError(t, err)
+			assert.Equal(t, must.NotFail(got.Get("iterationCount")), int32(15000))
+			assert.Len(t, must.NotFail(got.Get("salt")).(string), 60)
+			assert.Len(t, must.NotFail(got.Get("serverKey")).(string), 44)
+			assert.Len(t, must.NotFail(got.Get("storedKey")).(string), 44)
```
Comment (AlekSi) [inline internal/util/password/scramsha256_test.go:49]: The test should compare values, not their length

**Review by AlekSi — CHANGES_REQUESTED** [review]

File: internal/handler/msg_saslstart.go [hunk 11]
```diff
@@ -61,25 +65,39 @@ func (h *Handler) MsgSASLStart(ctx context.Context, msg *wire.OpMsg) (*wire.OpMs
...
+	case "SCRAM-SHA-256":
+		response, err := h.saslStartSCRAMSHA256(ctx, document)
+		if err != nil {
+			return nil, err
+		}
+
+		must.NoError(reply.SetSections(wire.MakeOpMsgSection(
+			must.NotFail(types.NewDocument(
+				"ok", float64(1),
+				"conversationId", int32(1),
```
[the remark this answers is not in the document]
Comment (b1ron) [reply internal/handler/msg_saslstart.go:89]: The `conversationId` doesn't seem to change for the duration of a conversation. See here for a [definition](https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#L431).

File: internal/handler/msg_saslstart.go [hunk 12]
```diff
@@ -61,25 +65,39 @@ func (h *Handler) MsgSASLStart(ctx context.Context, msg *wire.OpMsg) (*wire.OpMs
...
+				"done", false,
```
[the remark this answers is not in the document]
Comment (b1ron) [reply internal/handler/msg_saslstart.go:90]: This indicates that the conversation is not yet complete and that the client will send a proof in the subsequent `saslContinue` [command](https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#L651-L654).

[the remark this answers is not in the document]
Comment (henvic) [reply internal/handler/msg_saslstart.go:90]: Also, we don't have an explicit test case for this as it'd be hard to write such a test given that the `saslStart` command is triggered by the MongoDB client.

File: internal/handler/msg_saslstart.go [hunk 13]
```diff
@@ -129,3 +147,117 @@ func saslStartPlain(doc *types.Document) (string, string, error) {
...
+	}
+
+	return scram.StoredCredentials{}, errors.New("user not found")
+}
+
+func (h *Handler) saslStartSCRAMSHA256(ctx context.Context, doc *types.Document) (string, error) {
+	var payload []byte
+
+	binaryPayload, err := common.GetRequiredParam[types.Binary](doc, "payload")
+	if err == nil {
```
[the remark this answers is not in the document]
Comment (henvic) [reply internal/handler/msg_saslstart.go:239]: This wouldn't work. Fixing.

File: internal/util/password/scramsha256_test.go (hunk 10 above)
Comment (henvic) [reply internal/util/password/scramsha256_test.go:49]: The iterationCount is the only one I can compare value directly.
NewSCRAMSHA256 actually generates the other values:

* salt by calling crypto/rand.Read within NewSCRAMSHA256 itself
* serverKey and storedKey by calling `client.GetStoredCredentials`

Comment (AlekSi) [reply internal/util/password/scramsha256_test.go:49]: Yet existing code in that package _somehow_ solves that problem

File: internal/handler/msg_saslstart.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handler/msg_saslstart.go:89]: It seems strange to have a field that never changes. Perhaps, conversationId is used by other authentication mechanism and SCRAM-SHA-256 does not use it, I'm not sure how it works

File: internal/handler/msg_saslstart.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handler/msg_saslstart.go:90]: Thanks :pray:

File: internal/util/password/scramsha256.go (hunk 9 above)
Comment (b1ron) [reply internal/util/password/scramsha256.go:51]: It stores credentials as base64 encoded strings except for the iteration count. But there's a new problem see [thread](https://ferretdb-inc.slack.com/archives/C034LBM1J5V/p1706706734612669?thread_ts=1706704672.835109&cid=C034LBM1J5V) for details.

```
byron@byron-ThinkCentre-M910q:~/workdir/FerretDB$ mongo --quiet --port 47017
rs0:PRIMARY> use admin
switched to db admin
rs0:PRIMARY> typeof db.system.users.findOne().credentials['SCRAM-SHA-256'].storedKey
string
rs0:PRIMARY> db.system.users.findOne().credentials['SCRAM-SHA-256']
{
	"iterationCount" : 15000,
	"salt" : "2alXmLg2Yiz6XVWum5aYfW0OjqGrJfHukMSd1A==",
	"storedKey" : "y+bjwKtKYc66+5Tb+PZK5pngmsz/hsItkH0p7TPDg3g=",
	"serverKey" : "RRqDHT16fc82Kqykgbn4poCKfDkqOaDpPGx3QN9c0A0="
}
rs0:PRIMARY> 
```

File: internal/handler/msg_saslstart.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (b1ron) [reply internal/handler/msg_saslstart.go:89]: It seems to indicate some state that's remembered during the conversation, but it doesn't make a lot of sense to me either.

> conversationId: the conversation identifier. This will need to be remembered and used for the duration of the conversation.

File: internal/handler/msg_createuser.go [hunk 14]
```diff
@@ -218,3 +183,49 @@ func (h *Handler) MsgCreateUser(ctx context.Context, msg *wire.OpMsg) (*wire.OpM
 
 	return &reply, nil
 }
+
+// makeCredentials creates a document with credentials for the chosen mechanisms.
+func makeCredentials(mechanisms *types.Array, username, pwd string) (*types.Document, error) {
+	if mechanisms == nil {
```
[the remark this answers is not in the document]
Comment (henvic) [reply internal/handler/msg_createuser.go]: It's never nil. Removed this check.

File: internal/handler/msg_saslcontinue.go [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+// MsgSASLContinue implements `saslContinue` command.
+func (h *Handler) MsgSASLContinue(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, error) {
+	doc, err := msg.Document()
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	var payload []byte
+
+	// most drivers follow spec and send payload as a binary
```
Comment (chilagrow) [inline internal/handler/msg_saslcontinue.go:38]: This comment suggests there was a driver that sends payload in a different format (string?), do we want to keep allowing that?

File: internal/handler/handler.go [hunk 16]
```diff
@@ -192,7 +192,6 @@ func (h *Handler) cleanupAllCappedCollections(ctx context.Context) error {
 	}()
 
 	connInfo := conninfo.New()
-	connInfo.BypassBackendAuth = true
```
Comment (chilagrow) [inline internal/handler/handler.go:195]: Does cleaning capped collection work without this?

File: internal/handler/msg_saslstart.go [hunk 17]
```diff
@@ -129,3 +149,122 @@ func saslStartPlain(doc *types.Document) (string, string, error) {
 
 	return string(authcid), string(passwd), nil
 }
+
+func (h *Handler) credentialLookup(ctx context.Context, username, dbName string) (scram.StoredCredentials, error) {
```
Comment (chilagrow) [inline internal/handler/msg_saslstart.go:153]: Let's have comment for this function :hugs:

File: integration/users/connection_test.go [hunk 18]
```diff
@@ -165,15 +200,21 @@ func TestAuthentication(t *testing.T) {
...
-			require.NotEmpty(t, id)
-
-			var result bson.D
-			err = connCollection.FindOne(ctx, bson.D{{"_id", id}}).Decode(&result)
-			require.NoError(t, err, "cannot find document")
-			assert.Equal(t, bson.D{{"_id", id}, {"ping", "pong"}}, result)
+
+			require.NotNil(t, connCollection, "cannot get collection")
+
+			// TODO https://github.com/FerretDB/FerretDB/issues/3121
```
[the remark this answers is not in the document]
Comment (henvic) [reply integration/users/connection_test.go:206]: @chilagrow, yes. The problem is with the insertion. I discussed this with @b1ron once I tested this code and gotten an `unknown field "apiVersion"` error.

File: internal/handler/msg_saslstart.go (hunk 17 above)
Comment (henvic) [reply internal/handler/msg_saslstart.go:154]: Added.

File: internal/handler/handler.go (hunk 16 above)
Comment (henvic) [reply internal/handler/handler.go:195]: That was mistakenly removed and not readded. I added it back.

File: internal/handler/msg_saslcontinue.go (hunk 15 above)
Comment (henvic) [reply internal/handler/msg_saslcontinue.go]: Discussed this with @b1ron. He checked that no driver seems to be doing this for SCRAM and we decided to remove this comment for SCRAM. The alternative would be to [actually] add code to handle that:

```go
stringPayload, err := common.GetRequiredParam[string](doc, "payload")
if err == nil {
	if payload, err = base64.StdEncoding.DecodeString(stringPayload); err != nil {
		return "", "", handlererrors.NewCommandErrorMsgWithArgument(
			handlererrors.ErrBadValue,
			fmt.Sprintf("Invalid payload: %v", err),
			"payload",
		)
	}
}
```

File: internal/handler/msg_saslstart.go [hunk 19]
```diff
@@ -61,25 +65,41 @@ func (h *Handler) MsgSASLStart(ctx context.Context, msg *wire.OpMsg) (*wire.OpMs
...
+		}
+
+		conninfo.Get(ctx).BypassBackendAuth()
+
+		must.NoError(reply.SetSections(wire.MakeOpMsgSection(
+			must.NotFail(types.NewDocument(
+				"ok", float64(1),
+				"conversationId", int32(1),
+				"done", false,
+				"payload", response,
```
Comment (b1ron) [inline internal/handler/msg_saslstart.go:93]: I checkout the PR and was unable to authenticate because the payloads in the responses need to be binary. The tweak below allows a client to successfully authenticate with the exception of the legacy shell due to an invalid salt length and the dotnet driver because it uses the `skipEmptyExchange` option to avoid an extra round trip.

```go
	b := types.Binary{}
	b.B = []byte(resp)
	b.Subtype = 0

	var reply wire.OpMsg
	must.NoError(reply.SetSections(wire.MakeOpMsgSection(
		must.NotFail(types.NewDocument(
			"conversationId", int32(1),
			"done", true,
			"payload", b,
			"ok", float64(1),
		)),
	)))
```

With the exception of the legacy shell due to an invalid salt length and the dotnet driv

File: internal/handler/msg_saslcontinue.go [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+
+	resp, err := conv.Step(string(payload))
+	if err != nil {
+		return nil, handlererrors.NewCommandErrorMsg(
+			handlererrors.ErrAuthenticationFailed,
+			"Authentication failed.",
+		)
+	}
+
+	var reply wire.OpMsg
```
Comment (b1ron) [inline internal/handler/msg_saslcontinue.go:55]: Here too.

```go
	b := types.Binary{}
	b.B = []byte(resp)
	b.Subtype = 0

	var reply wire.OpMsg
	must.NoError(reply.SetSections(wire.MakeOpMsgSection(
		must.NotFail(types.NewDocument(
			"conversationId", int32(1),
			"done", true,
			"payload", b,
			"ok", float64(1),
		)),
	)))
```

File: internal/handler/msg_saslcontinue.go [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+	must.NoError(reply.SetSections(wire.MakeOpMsgSection(
+		must.NotFail(types.NewDocument(
+			"conversationId", int32(1),
+			"done", true,
+			"payload", resp,
```
Comment (b1ron) [inline internal/handler/msg_saslcontinue.go:60]: Here too, it needs to be of a `types.Binary`.

File: internal/handler/msg_saslstart.go (hunk 19 above)
Comment (b1ron) [reply internal/handler/msg_saslstart.go:93]: Just ignore what `skipEmptyExchange` means for now as it requires further investigation.

File: internal/handler/msg_saslcontinue.go (hunk 15 above)
Comment (chilagrow) [reply internal/handler/msg_saslcontinue.go]: Let's not add code. It's good news if all driver use binary :+1:

File: integration/users/connection_test.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (chilagrow) [reply integration/users/connection_test.go:206]: Ouch, `apiVersion` error seems to be a blocker for the new authentication code.

[the remark this answers is not in the document]
Comment (chilagrow) [reply integration/users/connection_test.go:206]: Insert will work if we don't set `.SetServerAPIOptions(...)` on client. Is there any reason we wanted to specify API version explicitly?

File: internal/handler/msg_saslstart.go (hunk 19 above)
Comment (chilagrow) [reply internal/handler/msg_saslstart.go]: Does this mean existing integration tests are not sufficient? :thinking: I mean manual testing is great, but I thought such scenario is covered by integration tests

File: integration/users/connection_test.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (henvic) [reply integration/users/connection_test.go:206]: @chilagrow, oh! Good point. Thanks! I've tried removing the server API version field and it worked flawlessly. I'm going to commit & push the change now.

File: internal/handler/msg_saslstart.go (hunk 19 above)
Comment (chilagrow) [reply internal/handler/msg_saslstart.go]: Yeah that's a nice to have.

File: internal/handler/msg_saslstart.go [hunk 22]
```diff
@@ -129,3 +153,124 @@ func saslStartPlain(doc *types.Document) (string, string, error) {
...
+				)
+			}
+
+			credentials := must.NotFail(v.Get("credentials")).(*types.Document)
+
+			if !credentials.Has("SCRAM-SHA-256") {
+				return scram.StoredCredentials{}, handlererrors.NewCommandErrorMsgWithArgument(
+					handlererrors.ErrMechanismUnavailable,
+					"User has no SCRAM-SHA-256 based authentication credentials registered",
+					"SCRAM-SHA-256",
```
Comment (chilagrow) [inline internal/handler/msg_saslstart.go:218]: Can we test this scenario?

**Review by chilagrow — DISMISSED** [review]
Overall looks great. See comment about 2 untested errors, which would be nice to test if possible.

File: internal/util/password/scramsha256.go [hunk 23]
```diff
@@ -52,7 +52,7 @@ type scramSHA256Params struct {
 // fixedScramSHA256Params represent fixed password parameters for SCRAM-SHA-256 authentication.
 var fixedScramSHA256Params = &scramSHA256Params{
 	iterationCount: 15_000,
-	saltLen:        45,
+	saltLen:        30,
```
Comment (henvic) [inline internal/util/password/scramsha256.go:55]: Adjusting to the same length used by MongoDB.

File: internal/handler/msg_saslstart.go [hunk 24]
```diff
@@ -129,3 +147,117 @@ func saslStartPlain(doc *types.Document) (string, string, error) {
...
+		}
+
+		matches, err := common.FilterDocument(v, filter)
+		if err != nil {
+			return scram.StoredCredentials{}, lazyerrors.Error(err)
+		}
+
+		if matches {
+			if !v.Has("credentials") {
+				return scram.StoredCredentials{}, handlererrors.NewCommandErrorMsgWithArgument(
```
[the remark this answers is not in the document]
Comment (henvic) [reply internal/handler/msg_saslstart.go:205]: It's not reachable given that credentials should always exist. I'm removing the check.

**Review by chilagrow — APPROVED** [review]

File: internal/handler/msg_usersinfo.go [hunk 25]
```diff
@@ -169,6 +169,18 @@ func (h *Handler) MsgUsersInfo(ctx context.Context, msg *wire.OpMsg) (*wire.OpMs
 			return nil, lazyerrors.Error(err)
 		}
 
+		if v.Has("credentials") {
+			credentials := must.NotFail(v.Get("credentials")).(*types.Document)
+			if credentialsKeys := credentials.Keys(); len(credentialsKeys) > 0 {
+				mechanisms := must.NotFail(types.NewArray())
+				for _, k := range credentialsKeys {
```
Comment (rumyantseva) [inline internal/handler/msg_usersinfo.go:176]: If `credentialsKeys` is a slice of strings, I guess we can do something like `types.NewArray(credentialsKeys...)` to save a few lines of code :)

File: internal/handler/msg_saslstart.go [hunk 26]
```diff
@@ -129,3 +153,116 @@ func saslStartPlain(doc *types.Document) (string, string, error) {
 
 	return string(authcid), string(passwd), nil
 }
+
+// scramCredentialLookup looks up an user's credentials in the database.
+func (h *Handler) scramCredentialLookup(ctx context.Context, username, dbName string) (scram.StoredCredentials, error) {
```
Comment (rumyantseva) [inline internal/handler/msg_saslstart.go:158]: Why do we return `scram.StoredCredentials` and not `*scram.StoredCredentials` in this case? 
(Personally, I'm ok with both option but the option with pointer looks a bit more typical for the rest of the code, and also semantically it's easier to interpret (nil pointer is returner in case of an error)

File: internal/handler/msg_saslstart.go [hunk 27]
```diff
@@ -129,3 +153,116 @@ func saslStartPlain(doc *types.Document) (string, string, error) {
...
+				StoredKey: storedKey,
+				ServerKey: serverKey,
+			}, nil
+		}
+	}
+
+	return scram.StoredCredentials{}, errors.New("user not found")
+}
+
+// saslStartSCRAMSHA256 extracts the initial challenge and respond the client.
```
Comment (rumyantseva) [inline internal/handler/msg_saslstart.go:234]: I don't quite understand the godoc line, so I suggest this (please check if it's correct):
```suggestion
// saslStartSCRAMSHA256 extracts the initial challenge, checks credentials, and returns the response that needs to be sent to the client.
//
// If credentials are not found, it returns XYZ.
```

File: internal/handler/msg_saslstart.go [hunk 28]
```diff
@@ -129,3 +153,116 @@ func saslStartPlain(doc *types.Document) (string, string, error) {
...
+					Salt:  string(salt),
+					Iters: int(must.NotFail(cred.Get("iterationCount")).(int32)),
+				},
+				StoredKey: storedKey,
+				ServerKey: serverKey,
+			}, nil
+		}
+	}
+
+	return scram.StoredCredentials{}, errors.New("user not found")
```
Comment (rumyantseva) [inline internal/handler/msg_saslstart.go:231]: This error message doesn't seem to be checked anywhere (in tests or in some parts of the code).

The other places in this function either return `lazyerror` or `handlererrors.NewCommandErrorMsg`. Should this one be a `lazyerror` or a `handlererrors.NewCommandErrorMsg` error too?

**Review by rumyantseva — DISMISSED** [review]
The code looks good. I added a few suggestions (mostly about incosistency), but they are not super crucial.

File: internal/handler/msg_usersinfo.go (hunk 25 above)
Comment (henvic) [reply internal/handler/msg_usersinfo.go:176]: I'm afraid we can't do that because types.NewArray receives a slice of []any and credentialsKeys is a slice of []string.

File: internal/handler/msg_saslstart.go (hunk 26 above)
Comment (henvic) [reply internal/handler/msg_saslstart.go:158]: I didn't do that because I was trying to avoid this weird `return *cred, nil` line here, so as it wasn't necessary I left it as this, but I see it's happening just on a single place, so I'm changing it now.

```go
scramServer, err := scram.SHA256.NewServer(func(username string) (scram.StoredCredentials, error) {
		cred, err := h.scramCredentialLookup(ctx, username, dbName, "SCRAM-SHA-256")
		if err != nil {
			return scram.StoredCredentials{}, err
		}
		return *cred, nil
	})
	if err != nil {
		return "", err
	}
```

File: internal/handler/msg_saslstart.go (hunk 27 above)
Comment (henvic) [reply internal/handler/msg_saslstart.go:234]: Probably

```
// saslStartSCRAMSHA256 extracts the initial challenge and attempts to move the
// authentication conversation forward.
```

(similar to how `*scram.ServerConversation.Step` explains) is more precise as it checks if the challenge can be resolved, rather than the credential itself.

File: internal/handler/msg_saslstart.go (hunk 19 above)
Comment (b1ron) [reply internal/handler/msg_saslstart.go]: Sorry nit: the payload variable is called `response` here, but `resp` in msg_saslcontinue.go.

File: internal/handler/msg_saslstart.go (hunk 26 above)
Comment (AlekSi) [reply internal/handler/msg_saslstart.go:158]: FWIW, it is our documented best practice: https://github.com/FerretDB/FerretDB/blob/e1eaaad891265ac947fbf55fd5f736ba301c9bea/CONTRIBUTING.md?plain=1#L260-L261

File: internal/clientconn/conninfo/conn_info.go [hunk 29]
```diff
@@ -35,12 +37,14 @@ type ConnInfo struct {
...
 
+	sc *scram.ServerConversation // protected by rw
+
 	// If true, backend implementations should not perform authentication
 	// by adding username and password to the connection string.
 	// It is set to true for background connections (such us capped collections cleanup)
 	// and by the new authentication mechanism.
 	// See where it is used for more details.
-	BypassBackendAuth bool
+	bypassBackendAuth bool
```
Comment (AlekSi) [inline internal/clientconn/conninfo/conn_info.go:47]:
```suggestion
	bypassBackendAuth bool // protected by rw
```

Normally, there would be a [mutex hat](https://dmitri.shuralyov.com/idiomatic-go/entries/2), but to make this struct smaller the order of fields is weird. So we should have a comment

File: internal/clientconn/conninfo/conn_info.go [hunk 30]
```diff
@@ -91,6 +112,22 @@ func (connInfo *ConnInfo) SetMetadataRecv() {
...
+	connInfo.bypassBackendAuth = true
+}
+
+// GetBypassBackendAuth returns whether the connection requires backend authentication.
+func (connInfo *ConnInfo) GetBypassBackendAuth() bool {
+	connInfo.rw.RLock()
+	defer connInfo.rw.RUnlock()
+
+	return connInfo.bypassBackendAuth
+}
```
Comment (AlekSi) [inline internal/clientconn/conninfo/conn_info.go:129]: Note how other methods are called `X` and `SetX`, not `GetX` and `X`.

File: internal/handler/msg_saslstart.go (hunk 28 above)
Comment (henvic) [reply internal/handler/msg_saslstart.go]: Fixed.

File: internal/clientconn/conninfo/conn_info.go (hunk 30 above)
Comment (henvic) [reply internal/clientconn/conninfo/conn_info.go]: Changed!

**Review by chilagrow — APPROVED** [review]

**Review by AlekSi — APPROVED** [review]
